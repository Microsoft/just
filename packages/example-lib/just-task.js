const { tscTask, task, parallel } = require('just-scripts');

task('typescript', tscTask({}));
task('typescript:watch', tscTask({ watch: true }));

task('build', parallel('typescript'));
task('watch', parallel('typescript:watch'));
