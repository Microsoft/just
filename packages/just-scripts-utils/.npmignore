src
CHANGELOG.*
__tests__