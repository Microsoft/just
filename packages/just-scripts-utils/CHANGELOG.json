{
  "name": "just-scripts-utils",
  "entries": [
    {
      "date": "Mon, 13 Apr 2020 15:57:30 GMT",
      "tag": "just-scripts-utils_v0.9.0",
      "version": "0.9.0",
      "comments": {
        "minor": [
          {
            "comment": "Add execSync utility",
            "author": "elcraig@microsoft.com",
            "commit": "91c6179993b4cc25076b8511b4d5d4726cd402d4"
          }
        ]
      }
    },
    {
      "date": "Tue, 13 Aug 2019 21:00:51 GMT",
      "tag": "just-scripts-utils_v0.8.4",
      "version": "0.8.4",
      "comments": {
        "patch": [
          {
            "comment": "Upgraded marked to 0.7.0 and marked-terminal to 3.3.0",
            "author": "altinokd@microsoft.com",
            "commit": "003c41cf6d3d3ac5bce48c6966475fcb6ceb7406"
          }
        ]
      }
    },
    {
      "date": "Mon, 05 Aug 2019 06:10:22 GMT",
      "tag": "just-scripts-utils_v0.8.3",
      "version": "0.8.3",
      "comments": {
        "patch": [
          {
            "comment": "Switches to using plop.js for codegen",
            "author": "kchau@microsoft.com",
            "commit": "2ea4e433c300ed8323aaab0fcf03f927ba8ab78b"
          }
        ]
      }
    },
    {
      "version": "0.8.1",
      "tag": "just-scripts-utils_v0.8.1",
      "date": "Sat, 01 Jun 2019 04:59:34 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Fix downloadPackage tests"
          }
        ]
      }
    },
    {
      "version": "0.8.0",
      "tag": "just-scripts-utils_v0.8.0",
      "date": "Fri, 31 May 2019 21:15:02 GMT",
      "comments": {
        "minor": [
          {
            "comment": "Adds the ability to override registry in downloadPackage"
          }
        ]
      }
    },
    {
      "version": "0.7.3",
      "tag": "just-scripts-utils_v0.7.3",
      "date": "Wed, 15 May 2019 18:44:04 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Switch microsoft in repo URL to lowercase"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"just-task-logger\" from `>=0.2.0 <1.0.0` to `>=0.2.1 <1.0.0`"
          }
        ]
      }
    },
    {
      "version": "0.7.2",
      "tag": "just-scripts-utils_v0.7.2",
      "date": "Thu, 25 Apr 2019 18:29:49 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Fix #78 by using Spawn instead of Exec"
          }
        ]
      }
    },
    {
      "version": "0.7.1",
      "tag": "just-scripts-utils_v0.7.1",
      "date": "Wed, 24 Apr 2019 23:38:23 GMT",
      "comments": {
        "patch": [
          {
            "comment": "up dep"
          },
          {
            "comment": "add ability to use path for template"
          }
        ]
      }
    },
    {
      "version": "0.7.0",
      "tag": "just-scripts-utils_v0.7.0",
      "date": "Wed, 10 Apr 2019 20:30:38 GMT",
      "comments": {
        "minor": [
          {
            "comment": "Fixing up the temp path to also use version in path"
          }
        ]
      }
    },
    {
      "version": "0.6.1",
      "tag": "just-scripts-utils_v0.6.1",
      "date": "Tue, 12 Mar 2019 20:07:18 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Exec error handling improvements"
          }
        ]
      }
    },
    {
      "version": "0.6.0",
      "tag": "just-scripts-utils_v0.6.0",
      "date": "Fri, 22 Feb 2019 23:37:36 GMT",
      "comments": {
        "minor": [
          {
            "comment": "Consistent interface naming"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"just-task-logger\" from `>=0.1.2 <1.0.0` to `>=0.2.0 <1.0.0`"
          }
        ]
      }
    },
    {
      "version": "0.5.0",
      "tag": "just-scripts-utils_v0.5.0",
      "date": "Fri, 22 Feb 2019 22:50:01 GMT",
      "comments": {
        "minor": [
          {
            "comment": "Move exec to just-scripts-utils"
          }
        ]
      }
    },
    {
      "version": "0.4.2",
      "tag": "just-scripts-utils_v0.4.2",
      "date": "Fri, 22 Feb 2019 19:03:38 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Update package metadata"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"just-task-logger\" from `>=0.1.1 <1.0.0` to `>=0.1.2 <1.0.0`"
          }
        ]
      }
    },
    {
      "version": "0.4.1",
      "tag": "just-scripts-utils_v0.4.1",
      "date": "Thu, 21 Feb 2019 23:25:53 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Use new logger package"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"just-task-logger\" from `>=0.1.0 <1.0.0` to `>=0.1.1 <1.0.0`"
          }
        ]
      }
    },
    {
      "version": "0.4.0",
      "tag": "just-scripts-utils_v0.4.0",
      "date": "Wed, 13 Feb 2019 22:31:47 GMT",
      "comments": {
        "minor": [
          {
            "comment": "API updates and error checking"
          }
        ]
      }
    },
    {
      "version": "0.3.3",
      "tag": "just-scripts-utils_v0.3.3",
      "date": "Mon, 04 Feb 2019 05:47:04 GMT",
      "comments": {
        "patch": [
          {
            "comment": "flexibility change"
          }
        ]
      }
    },
    {
      "version": "0.3.2",
      "tag": "just-scripts-utils_v0.3.2",
      "date": "Fri, 25 Jan 2019 07:08:59 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Fixes"
          }
        ]
      }
    },
    {
      "version": "0.3.1",
      "tag": "just-scripts-utils_v0.3.1",
      "date": "Wed, 23 Jan 2019 00:05:38 GMT",
      "comments": {
        "patch": [
          {
            "comment": "info as it is adding package"
          }
        ]
      }
    },
    {
      "version": "0.3.0",
      "tag": "just-scripts-utils_v0.3.0",
      "date": "Tue, 22 Jan 2019 22:31:23 GMT",
      "comments": {
        "minor": [
          {
            "comment": "move utils from scripts to here"
          }
        ]
      }
    },
    {
      "version": "0.2.2",
      "tag": "just-scripts-utils_v0.2.2",
      "date": "Mon, 21 Jan 2019 03:27:17 GMT",
      "comments": {
        "patch": [
          {
            "comment": "fix scripts to transform correctly with subdirs"
          }
        ]
      }
    },
    {
      "version": "0.2.1",
      "tag": "just-scripts-utils_v0.2.1",
      "date": "Mon, 21 Jan 2019 02:54:15 GMT",
      "comments": {
        "patch": [
          {
            "comment": "fix upgrade stack task"
          }
        ]
      }
    },
    {
      "version": "0.2.0",
      "tag": "just-scripts-utils_v0.2.0",
      "date": "Mon, 21 Jan 2019 02:21:01 GMT",
      "comments": {
        "minor": [
          {
            "comment": "new publish!"
          }
        ]
      }
    }
  ]
}