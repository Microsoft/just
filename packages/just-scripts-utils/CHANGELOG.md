# Changelog - just-scripts-utils

This log was last generated on Mon, 01 Jul 2019 20:15:20 GMT and should not be manually modified.

## 0.8.2
Mon, 01 Jul 2019 20:15:20 GMT

### Patches

- introduce ESLint to sources (43081j@users.noreply.github.com)

## 0.8.1
Wed, 12 Jun 2019 02:03:29 GMT

## 0.8.1
Wed, 12 Jun 2019 01:49:55 GMT

## 0.8.1
Sat, 01 Jun 2019 04:59:34 GMT

### Patches

- Fix downloadPackage tests

## 0.8.0
Fri, 31 May 2019 21:15:02 GMT

### Minor changes

- Adds the ability to override registry in downloadPackage

## 0.7.3
Wed, 15 May 2019 18:44:04 GMT

### Patches

- Switch microsoft in repo URL to lowercase

## 0.7.2
Thu, 25 Apr 2019 18:29:49 GMT

### Patches

- Fix #78 by using Spawn instead of Exec

## 0.7.1
Wed, 24 Apr 2019 23:38:23 GMT

### Patches

- up dep
- add ability to use path for template

## 0.7.0
Wed, 10 Apr 2019 20:30:38 GMT

### Minor changes

- Fixing up the temp path to also use version in path

## 0.6.1
Tue, 12 Mar 2019 20:07:18 GMT

### Patches

- Exec error handling improvements

## 0.6.0
Fri, 22 Feb 2019 23:37:36 GMT

### Minor changes

- Consistent interface naming

## 0.5.0
Fri, 22 Feb 2019 22:50:01 GMT

### Minor changes

- Move exec to just-scripts-utils

## 0.4.2
Fri, 22 Feb 2019 19:03:38 GMT

### Patches

- Update package metadata

## 0.4.1
Thu, 21 Feb 2019 23:25:53 GMT

### Patches

- Use new logger package

## 0.4.0
Wed, 13 Feb 2019 22:31:47 GMT

### Minor changes

- API updates and error checking

## 0.3.3
Mon, 04 Feb 2019 05:47:04 GMT

### Patches

- flexibility change

## 0.3.2
Fri, 25 Jan 2019 07:08:59 GMT

### Patches

- Fixes

## 0.3.1
Wed, 23 Jan 2019 00:05:38 GMT

### Patches

- info as it is adding package

## 0.3.0
Tue, 22 Jan 2019 22:31:23 GMT

### Minor changes

- move utils from scripts to here

## 0.2.2
Mon, 21 Jan 2019 03:27:17 GMT

### Patches

- fix scripts to transform correctly with subdirs

## 0.2.1
Mon, 21 Jan 2019 02:54:15 GMT

### Patches

- fix upgrade stack task

## 0.2.0
Mon, 21 Jan 2019 02:21:01 GMT

### Minor changes

- new publish!

