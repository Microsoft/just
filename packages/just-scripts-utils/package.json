{
  "name": "just-scripts-utils",
  "version": "0.3.3",
  "description": "Utilities for Just stack scripts",
  "main": "./lib/index.js",
  "scripts": {
    "build": "tsc",
    "start": "tsc -w --preserveWatchOutput",
    "test": "jest",
    "start-test": "jest --watch"
  },
  "keywords": [],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "chalk": "^2.4.1",
    "fs-extra": "^7.0.1",
    "glob": "^7.1.3",
    "handlebars": "^4.0.12",
    "jju": "^1.4.0",
    "marked": "^0.6.0",
    "marked-terminal": "^3.2.0",
    "semver": "^5.6.0",
    "tar": "^4.4.8",
    "yargs": "^12.0.5"
  },
  "devDependencies": {
    "@types/fs-extra": "^5.0.4",
    "@types/glob": "^7.1.1",
    "@types/handlebars": "^4.0.39",
    "@types/jest": "^23.3.13",
    "@types/jju": "^1.4.0",
    "@types/marked": "^0.6.0",
    "@types/marked-terminal": "^3.1.1",
    "@types/mock-fs": "^3.6.30",
    "@types/node": "^10.12.18",
    "@types/semver": "^5.5.0",
    "@types/tar": "^4.0.0",
    "@types/yargs": "12.0.1",
    "jest": "^23.6.0",
    "mock-fs": "^4.8.0",
    "ts-jest": "^23.10.5"
  }
}
