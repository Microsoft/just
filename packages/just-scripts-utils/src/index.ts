export { downloadPackage } from './downloadPackage';
export * from './findMonoRepoRootPath';
export * from './IPackageJson';
export * from './IRushJson';
export { logger } from './logger';
export * from './mergePackageJson';
export * from './paths';
export * from './prettyPrintMarkdown';
export * from './readPackageJson';
export { readRushJson, rushAddPackage, rushUpdate } from './rush';
export * from './transform';
