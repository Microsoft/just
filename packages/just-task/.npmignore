src
CHANGELOG.*
*.tgz
*.log
package-deps.json
jest.config.js
tsconfig.json