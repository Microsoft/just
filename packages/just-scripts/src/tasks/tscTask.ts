import ts from 'typescript';
import { resolve, logger, resolveCwd, TaskFunction } from 'just-task';
import { exec, encodeArgs, spawn } from 'just-scripts-utils';
import fs from 'fs';

export type TscTaskOptions = { [key in keyof ts.CompilerOptions]?: string | boolean };

export function tscTask(options: TscTaskOptions): TaskFunction {
  const tscCmd = resolve('typescript/lib/tsc.js');

  if (!tscCmd) {
    throw new Error('cannot find tsc');
  }

  return function tsc() {
    // Read from options argument, if not there try the tsConfigFile found in root, if not then skip and use no config
    options = { ...options, ...getProjectOrBuildOptions(options) };

    if (isValidProject(options)) {
      logger.info(`Running ${tscCmd} with ${options.project || options.build}`);

      const args = Object.keys(options).reduce(
        (args, option) => {
          if (typeof options[option] === 'string') {
            return args.concat(['--' + option, options[option] as string]);
          } else if (typeof options[option] === 'boolean') {
            return args.concat(['--' + option]);
          }

          return args;
        },
        [tscCmd]
      );

      const cmd = encodeArgs([process.execPath, ...args]).join(' ');
      logger.info(`Executing: ${cmd}`);
      return exec(cmd);
    } else {
      Promise.resolve();
    }

    return;
  };
}

export function tscWatchTask(options: TscTaskOptions): TaskFunction {
  const tscCmd = resolve('typescript/lib/tsc.js');

  if (!tscCmd) {
    throw new Error('cannot find tsc');
  }

  options = { ...options, ...getProjectOrBuildOptions(options) };

  return function tscWatch() {
    if (isValidProject(options)) {
      logger.info(`Running ${tscCmd} with ${options.project || options.build} in watch mode`);

      const args = Object.keys(options).reduce(
        (args, option) => {
          if (typeof options[option] === 'string') {
            return args.concat(['--' + option, options[option] as string]);
          } else if (typeof options[option] === 'boolean') {
            return args.concat(['--' + option]);
          }

          return args;
        },
        [tscCmd]
      );

      const cmd = [...args, '--watch'];
      logger.info(encodeArgs(cmd).join(' '));
      return spawn(process.execPath, cmd, { stdio: 'inherit' });
    } else {
      Promise.resolve();
    }

    return;
  };
}

function getProjectOrBuildOptions(options: TscTaskOptions) {
  const tsConfigFile = resolveCwd('./tsconfig.json') || 'tsconfig.json';
  const result: { [option: string]: string | boolean } = {};

  if (options.project) {
    result.project = options && typeof options.project === 'string' ? options.project : tsConfigFile;
  } else if (options.build) {
    result.build = options && typeof options.build === 'string' ? options.build : tsConfigFile;
  } else if (tsConfigFile) {
    result.project = tsConfigFile;
  }

  return result;
}

function isValidProject(options: TscTaskOptions) {
  return (options.project && fs.existsSync(options.project as string)) || (options.build && fs.existsSync(options.build as string));
}
