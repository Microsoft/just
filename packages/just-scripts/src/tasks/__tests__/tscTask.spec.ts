import mockfs from 'mock-fs';
import { encodeArgs, exec, spawn } from 'just-scripts-utils';
import { TaskFunction } from 'just-task';
import { tscTask, tscWatchTask, TscTaskOptions } from '../tscTask';
import { callTaskForTest } from './callTaskForTest';
import { normalizeCmdArgsForTest } from './normalizeCmdArgsForTest';

// Jest will hoist these before the imports above, so these modules will be mocked first
jest.mock('just-scripts-utils/lib/exec', () => {
  const { mockExecFactory } = require('./mockExecFactory');
  return mockExecFactory();
});
jest.mock('just-task/lib/logger');

/**
 * Returns the composition of the `tsc.js` Node module in terms `mock-fs` understands, which is necessary for Node's
 * module loader to succeed.
 */
function mockFsTsc(relativePath?: string) {
  // Relative to cwd when the test runs
  const relativeRepoRoot = '../..';
  const ourRelativePath = relativePath || relativeRepoRoot;

  const mockFsConfig: { [key: string]: any } = {};
  mockFsConfig[`${ourRelativePath}/node_modules/typescript/lib/tsc.js`] = 'a file';
  mockFsConfig[`${ourRelativePath}/node_modules/typescript/package.json`] = 'a file';
  return mockFsConfig;
}

interface Given {
  tscTaskFn: (options: TscTaskOptions) => TaskFunction;
}

interface Expected {
  execOrSpawnSpy: jest.Mock<any>;
}

type TaskTestCase = [/* name */ string, Given, Expected];

describe(`tscTask`, () => {
  afterEach(() => {
    jest.clearAllMocks();
    mockfs.restore();
  });

  /**
   * Both `tscTask` and `tscWatchTask` should handle test cases similarly.
   */
  describe.each<TaskTestCase>([
    ['tscTask', { tscTaskFn: tscTask }, { execOrSpawnSpy: exec as jest.Mock<any> }],
    ['tscWatchTask', { tscTaskFn: tscWatchTask }, { execOrSpawnSpy: spawn as jest.Mock<any> }]
  ])(`using '%s' function`, (_name, given, expected) => {
    describe(`with empty options and tsconfig.json exists at package root`, () => {
      it(`execs expected command`, () => {
        mockfs({
          ...mockFsTsc(),
          'tsconfig.json': 'a file'
        });
        const task = given.tscTaskFn({});
        expect.assertions(3);
        return callTaskForTest(task).then(() => {
          // Restore mockfs so snapshots work
          mockfs.restore();
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
          // Inspect the call to `encodeArgs` since it is easier to strip out repo-specific path values.
          expect(encodeArgs).toHaveBeenCalled();
          const actualCmdArgs = normalizeCmdArgsForTest((encodeArgs as jest.Mock<any>).mock.calls[0][0]);
          expect(actualCmdArgs).toMatchSnapshot();
        });
      });
    });

    describe(`with empty options and tsconfig.json does not exist at package root`, () => {
      it(`does not exec command`, () => {
        mockfs({
          ...mockFsTsc()
        });
        const task = given.tscTaskFn({});
        expect.assertions(1);
        return callTaskForTest(task).then(() => {
          expect(expected.execOrSpawnSpy).not.toHaveBeenCalled();
        });
      });
    });

    describe(`with 'project' option where 'project' is custom path and tsconfig.json exists`, () => {
      it(`execs expected command`, () => {
        mockfs({
          ...mockFsTsc(),
          'a/custom/path': {
            'tsconfig.json': 'a file'
          }
        });
        const task = given.tscTaskFn({ project: 'a/custom/path/tsconfig.json' });
        expect.assertions(3);
        return callTaskForTest(task).then(() => {
          // Restore mockfs so snapshots work
          mockfs.restore();
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
          // Inspect the call to `encodeArgs` since it is easier to strip out repo-specific path values.
          expect(encodeArgs).toHaveBeenCalled();
          const actualCmdArgs = normalizeCmdArgsForTest((encodeArgs as jest.Mock<any>).mock.calls[0][0]);
          expect(actualCmdArgs).toMatchSnapshot();
        });
      });
    });

    describe(`with 'project' option where 'project' is custom path and tsconfig.json does not exist`, () => {
      it(`does not exec command`, () => {
        mockfs({
          ...mockFsTsc()
        });
        const task = given.tscTaskFn({ project: 'a/custom/path/tsconfig.json' });
        expect.assertions(1);
        return callTaskForTest(task).then(() => {
          expect(expected.execOrSpawnSpy).not.toHaveBeenCalled();
        });
      });
    });

    describe(`with 'build' option where 'build' is custom path and tsconfig.json exists`, () => {
      it(`execs expected command`, () => {
        mockfs({
          ...mockFsTsc(),
          'a/custom/path': {
            'tsconfig.json': 'a file'
          }
        });
        const task = given.tscTaskFn({ build: 'a/custom/path/tsconfig.json' });
        expect.assertions(3);
        return callTaskForTest(task).then(() => {
          // Restore mockfs so snapshots work
          mockfs.restore();
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
          // Inspect the call to `encodeArgs` since it is easier to strip out repo-specific path values.
          expect(encodeArgs).toHaveBeenCalled();
          const actualCmdArgs = normalizeCmdArgsForTest((encodeArgs as jest.Mock<any>).mock.calls[0][0]);
          expect(actualCmdArgs).toMatchSnapshot();
        });
      });
    });

    describe(`with 'build' option where 'build' is custom path and tsconfig.json does not exist`, () => {
      it(`does not exec command`, () => {
        mockfs({
          ...mockFsTsc()
        });
        const task = given.tscTaskFn({ build: 'a/custom/path/tsconfig.json' });
        expect.assertions(1);
        return callTaskForTest(task).then(() => {
          expect(expected.execOrSpawnSpy).not.toHaveBeenCalled();
        });
      });
    });

    describe(`with a boolean switch`, () => {
      it(`execs expected command`, () => {
        mockfs({
          ...mockFsTsc(),
          'tsconfig.json': 'a file'
        });
        const task = given.tscTaskFn({ allowJs: true });
        expect.assertions(3);
        return callTaskForTest(task).then(() => {
          // Restore mockfs so snapshots work
          mockfs.restore();
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
          // Inspect the call to `encodeArgs` since it is easier to strip out repo-specific path values.
          expect(encodeArgs).toHaveBeenCalled();
          const actualCmdArgs = normalizeCmdArgsForTest((encodeArgs as jest.Mock<any>).mock.calls[0][0]);
          expect(actualCmdArgs).toMatchSnapshot();
        });
      });
    });

    describe(`with a combination of switches`, () => {
      it(`execs expected command`, () => {
        mockfs({
          ...mockFsTsc(),
          'tsconfig.json': 'a file'
        });
        const task = given.tscTaskFn({ allowJs: true, build: 'tsconfig.json', outDir: 'some/out/path' });
        expect.assertions(3);
        return callTaskForTest(task).then(() => {
          // Restore mockfs so snapshots work
          mockfs.restore();
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
          // Inspect the call to `encodeArgs` since it is easier to strip out repo-specific path values.
          expect(encodeArgs).toHaveBeenCalled();
          const actualCmdArgs = normalizeCmdArgsForTest((encodeArgs as jest.Mock<any>).mock.calls[0][0]);
          expect(actualCmdArgs).toMatchSnapshot();
        });
      });
    });

    describe(`where repo has TypeScript installed`, () => {
      it(`execs command`, () => {
        mockfs({
          ...mockFsTsc(),
          'tsconfig.json': 'a file'
        });
        const task = given.tscTaskFn({});
        expect.assertions(1);
        return callTaskForTest(task).then(() => {
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
        });
      });
    });

    describe(`where package has TypeScript installed`, () => {
      it(`execs command`, () => {
        mockfs({
          ...mockFsTsc('.'),
          'tsconfig.json': 'a file'
        });
        const task = given.tscTaskFn({});
        expect.assertions(1);
        return callTaskForTest(task).then(() => {
          expect(expected.execOrSpawnSpy).toHaveBeenCalled();
        });
      });
    });

    describe(`where repo and package do not have TypeScript installed`, () => {
      it(`returns error`, () => {
        mockfs({
          'tsconfig.json': 'a file'
        });
        expect.assertions(1);
        expect(() => {
          given.tscTaskFn({});
        }).toThrow('cannot find tsc');
      });
    });
  });
});
