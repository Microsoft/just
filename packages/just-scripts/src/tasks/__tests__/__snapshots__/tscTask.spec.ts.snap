// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tscTask using 'tscTask' function with 'build' option where 'build' is custom path and tsconfig.json exists execs expected command 1`] = `
Array [
  "\${nodeExePath}",
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--build",
  "a/custom/path/tsconfig.json",
]
`;

exports[`tscTask using 'tscTask' function with 'project' option where 'project' is custom path and tsconfig.json exists execs expected command 1`] = `
Array [
  "\${nodeExePath}",
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--project",
  "a/custom/path/tsconfig.json",
]
`;

exports[`tscTask using 'tscTask' function with a boolean switch execs expected command 1`] = `
Array [
  "\${nodeExePath}",
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--allowJs",
  "--project",
  "\${packageRoot}/tsconfig.json",
]
`;

exports[`tscTask using 'tscTask' function with a combination of switches execs expected command 1`] = `
Array [
  "\${nodeExePath}",
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--allowJs",
  "--build",
  "tsconfig.json",
  "--outDir",
  "some/out/path",
]
`;

exports[`tscTask using 'tscTask' function with empty options and tsconfig.json exists at package root execs expected command 1`] = `
Array [
  "\${nodeExePath}",
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--project",
  "\${packageRoot}/tsconfig.json",
]
`;

exports[`tscTask using 'tscWatchTask' function with 'build' option where 'build' is custom path and tsconfig.json exists execs expected command 1`] = `
Array [
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--build",
  "a/custom/path/tsconfig.json",
  "--watch",
]
`;

exports[`tscTask using 'tscWatchTask' function with 'project' option where 'project' is custom path and tsconfig.json exists execs expected command 1`] = `
Array [
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--project",
  "a/custom/path/tsconfig.json",
  "--watch",
]
`;

exports[`tscTask using 'tscWatchTask' function with a boolean switch execs expected command 1`] = `
Array [
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--allowJs",
  "--project",
  "\${packageRoot}/tsconfig.json",
  "--watch",
]
`;

exports[`tscTask using 'tscWatchTask' function with a combination of switches execs expected command 1`] = `
Array [
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--allowJs",
  "--build",
  "tsconfig.json",
  "--outDir",
  "some/out/path",
  "--watch",
]
`;

exports[`tscTask using 'tscWatchTask' function with empty options and tsconfig.json exists at package root execs expected command 1`] = `
Array [
  "\${repoRoot}/node_modules/typescript/lib/tsc.js",
  "--project",
  "\${packageRoot}/tsconfig.json",
  "--watch",
]
`;
